File: app/src/main/java/io/ipoli/android/quest/data/Recurrence.java
package io.ipoli.android.quest.data;

import com.fasterxml.jackson.annotation.JsonIgnore;

import org.threeten.bp.LocalDate;

import io.ipoli.android.app.utils.DateUtils;

/**
 * Created by Venelin Valkov <venelin@curiousily.com>
 * on 4/8/16.
 */
public class Recurrence {

    public static final String RRULE_EVERY_DAY = "FREQ=WEEKLY;BYDAY=MO,TU,WE,TH,FR,SA,SU";

    public enum RepeatType {DAILY, WEEKLY, MONTHLY, YEARLY;}

    private int flexibleCount;

    private String rrule;

    private String rdate;

    private String exrule;
    private String exdate;
    private Long dtstart;
    private Long dtend;
    private String repeatType;

    public Recurrence() {

    }

    public static Recurrence create() {
        Recurrence recurrence = new Recurrence();
        recurrence.setRecurrenceType(RepeatType.DAILY);
        recurrence.setDtstartDate(LocalDate.now());
        recurrence.setFlexibleCount(0);
        return recurrence;
    }

    @JsonIgnore
    public void setRecurrenceType(RepeatType type) {
        this.repeatType = type.name();
    }

    public void setRepeatType(String repeatType) {
        this.repeatType = repeatType;
    }

    public String getRepeatType() {
        return repeatType;
    }

    public String getRrule() {
        return rrule;
    }

    public void setRrule(String rrule) {
        this.rrule = rrule;
    }

    public String getRdate() {
        return rdate;
    }

    public void setRdate(String rdate) {
        this.rdate = rdate;
    }

    public String getExrule() {
        return exrule;
    }

    public void setExrule(String exrule) {
        this.exrule = exrule;
    }

    public String getExdate() {
        return exdate;
    }

    public void setExdate(String exdate) {
        this.exdate = exdate;
    }

    @JsonIgnore
    public LocalDate getDtstartDate() {
        return dtstart != null ? DateUtils.fromMillis(dtstart) : null;
    }

    @JsonIgnore
    public void setDtstartDate(LocalDate dtstartDate) {
        dtstart = dtstartDate != null ? DateUtils.toMillis(dtstartDate) : null;
    }

    public Long getDtstart() {
        return dtstart;
    }

    public void setDtstart(Long dtstart) {
        this.dtstart = dtstart;
    }

    @JsonIgnore
    public LocalDate getDtendDate() {
        return dtend != null ? DateUtils.fromMillis(dtend) : null;
    }

    @JsonIgnore
    public void setDtendDate(LocalDate dtendDate) {
        dtend = dtendDate != null ? DateUtils.toMillis(dtendDate) : null;
    }

    public Long getDtend() {
        return dtend;
    }

    public void setDtend(Long dtend) {
        this.dtend = dtend;
    }

    public int getFlexibleCount() {
        return flexibleCount;
    }

    public void setFlexibleCount(int flexibleCount) {
        this.flexibleCount = flexibleCount;
    }

    @JsonIgnore
    public RepeatType getRecurrenceType() {
        return RepeatType.valueOf(repeatType);
    }

    @JsonIgnore
    public boolean isFlexible() {
        return flexibleCount > 0;
    }
}

File: app/src/main/java/io/ipoli/android/repeatingquest/data/Recurrence.kt
package io.ipoli.android.repeatingquest.data

import io.ipoli.android.common.datetime.DateUtils
import io.realm.RealmObject
import org.threeten.bp.LocalDate

/**
 * Created by Venelin Valkov venelin@curiousily.com>
 * on 8/22/17.
 */
open class Recurrence : RealmObject() {

    enum class RepeatType {
        DAILY, WEEKLY, MONTHLY, YEARLY
    }

    var flexibleCount: Int = 0

    var rrule: String? = null

    var rdate: String? = null

    var exrule: String? = null
    var exdate: String? = null
    var dtstart: Long? = null
    var dtend: Long? = null
    var repeatType: String? = null

    var dtstartDate: LocalDate?
        get() = if (dtstart != null) DateUtils.fromMillis(dtstart!!) else null
        set(dtstartDate) {
            dtstart = if (dtstartDate != null) DateUtils.toMillis(dtstartDate) else null
        }

    var dtendDate: LocalDate?
        get() = if (dtend != null) DateUtils.fromMillis(dtend!!) else null
        set(dtendDate) {
            dtend = if (dtendDate != null) DateUtils.toMillis(dtendDate) else null
        }

    var recurrenceType: RepeatType
        get() = RepeatType.valueOf(repeatType!!)
        set(type) {
            this.repeatType = type.name
        }

    val isFlexible: Boolean
        get() = flexibleCount > 0

    companion object {

        val RRULE_EVERY_DAY = "FREQ=WEEKLY;BYDAY=MO,TU,WE,TH,FR,SA,SU"

        fun create(): Recurrence {
            val recurrence = Recurrence()
            recurrence.recurrenceType = RepeatType.DAILY
            recurrence.dtstartDate = LocalDate.now()
            recurrence.flexibleCount = 0
            return recurrence
        }
    }
}